sdk/core/azure-core-serializer-avro-apache/src/test/java/com/azure/core/serializer/avro/apache/ApacheAvroSerializerTests.java

```diff
@@ -54,61 +55,56 @@ private static ApacheAvroSerializer getSerializer(String schema) {
 
     @ParameterizedTest
     @MethodSource("deserializePrimitiveTypesSupplier")
-    public <T> void deserializePrimitiveTypes(InputStream avro, String schema, Class<T> clazz, T expected) {
-        StepVerifier.create(getSerializer(schema).deserialize(avro, clazz))
-            .assertNext(actual -> {
-                assertTrue(clazz.isAssignableFrom(actual.getClass()));
-                assertEquals(expected, actual);
-            })
+    public <T> void deserializePrimitiveTypes(InputStream avro, String schema, TypeReference<T> type, T expected) {
+        StepVerifier.create(getSerializer(schema).deserialize(avro, type))
+            .assertNext(actual -> assertEquals(expected, actual))
             .verifyComplete();
     }
 
     private static Stream<Arguments> deserializePrimitiveTypesSupplier() {
         return Stream.of(
-            Arguments.of(streamCreator(0), schemaCreator("boolean"), Boolean.class, false),
-            Arguments.of(streamCreator(1), schemaCreator("boolean"), Boolean.class, true),
+            Arguments.of(streamCreator(0), schemaCreator("boolean"), new TypeReference<Boolean>() { }, false),
+            Arguments.of(streamCreator(1), schemaCreator("boolean"), new TypeReference<Boolean>() { }, true),
```

Comment:
Can we avoid creating a new instance for same type here and also have a way to minimize duplicate instances when used by the user? Maybe have a static map of known types and vend those `TypeReference` instances if one exists or create one if it doesnt?

Reply:
Basically create a backing map of static references for common built-in Java types such as `Boolean`, `Integer`, `Map<String, Object>`, etc?

Reply:
Yeah, basically, support these built-in types out of the box and for custom types, lazily add to the backing map.

Reply:
The best I can think of would be an API such as this.

```java
<T> TypeReference<T> createInstance(Class<T> clazz);
```

With a static cache of `Map<Class<T>, TypeReference<T>>`. This won't be able to handle `ParameterizedType`.

Is this something we would need to GA this feature or could we add it later?

Reply:
Adding this after GA would lead to 2 different ways of creating an instance of `TypeReference` - ctor and static method. So, we have to finalize this before GA.

Reply:
I would stick with using constructor as this matches patterns used in other similar concepts.